src/System.Management.Automation/engine/debugger/debugger.cs

```diff
@@ -2634,6 +2795,96 @@ internal override void StopDebugRunspace(Runspace runspace)
 
         #endregion
 
+        #region Runspace Debug Processing
+
+        /// <summary>
+        /// Adds the provided Runspace object to the runspace debugger processing queue.
+        /// The queue will then raise the ProcessRunspaceDebug events for each runspace to allow
+        /// a host script debugger implementation to provide an active debugging session.
+        /// </summary>
+        /// <param name="runspace">Runspace to debug</param>
+        internal override void QueueRunspaceForDebug(Runspace runspace)
+        {
+            if (runspace == null) { throw new PSArgumentNullException("runspace"); }
+
+            runspace.StateChanged += RunspaceStateChangedHandler;
+            runspace.AvailabilityChanged += RunspaceAvailabilityChangedHandler;
+            _runspaceDebugQueue.Value.Enqueue(new ProcessRunspaceDebugEventArgs(runspace));
+            StartRunspaceForDebugQueueProcessing();
+        }
+
+        /// <summary>
+        /// Causes the CancelRunspaceDebugProcessing event to be raised which notifies subscribers that these debugging
+        /// sessions should be cancelled.
+        /// </summary>
+        public override void CancelDebuggerProcessing()
+        {
+            try
+            {
+                RaiseCancelProcessRunspaceDebugEvent();
+            }
+            catch (Exception)
+            { }
+
+            ReleaseInternalRunspaceDebugProcessing(null, true);
```

Comment:
Actually I think it makes more sense to raise the event after internal clean up.